spec/lib/mercury/mercury_spec.rb

```diff
@@ -75,12 +75,85 @@
     end
   end
 
-  def with_mercury(&block)
+  # Commented out until this gets fixed: https://github.com/eventmachine/eventmachine/issues/670
+  # it 'raises an error when a connection cannot be established' do
+  #   expect { em { Mercury.open(port: 31999) { done } } }.to raise_error /Failed to establish connection/
+  #   expect(EM.reactor_running?).to be false
+  # end
+
+  it 'raises an error when the connection breaks' do
+    expect { em { Mercury.open { done } } }.to raise_error /Lost connection/
+    expect(EM.reactor_running?).to be false   # make sure we're not triggering EventMachine cleanup bugs
+  end
+
+  it 'does not obscure exceptions thrown inside the reactor' do
+    expect { em { Mercury.open { raise 'oops' } } }.to raise_error 'oops'
+    expect(EM.reactor_running?).to be false   # make sure we're not triggering EventMachine cleanup bugs
+  end
+
+  describe '#publish' do
+    context 'docker assumptions' do
+      after(:each) { start_rabbitmq_server }
+      # This test is commented out because it make assumptions about, and manipulates, a local docker orchestration.
+      # Uncomment and run it in a suitable environment to do semi-automated testing on publisher confirms.
+      # it 'waits to invoke its continuation until after the message is confirmed' do
+      #   got_confirmation = false
+      #   expect {
+      #     with_mercury(timeout_seconds: 15) do |m|
+      #       m.publish(source, 'hello') do # cause `source` to be declared so subsequent publishes simply publish
+      #         stop_rabbitmq_server
+      #         m.publish(source, 'hello') do
+      #           got_confirmation = true
+      #           m.close { done }
+      #         end
+      #       end
+      #     end
+      #   }.to raise_error /Lost connection to AMQP server/
+      #   expect(got_confirmation).to be false
+      # end
+    end
+    it 'allows multiple outstanding confirmations' do
+      log = []
+      with_mercury do |m|
+        m.publish(source, 'hello') do # cause `source` to be declared so subsequent publishes simply publish
+          publish_and_confirm(m, 'a', log)
+          publish_and_confirm(m, 'b', log)
+          em_wait_until(proc { log.size == 4 }) do
+            expect(log).to eql ['publish a', 'publish b', 'confirm a', 'confirm b']
```

Comment:
The _server_ may very well confirm `a` before it receives `b`, but the _client_ won't observe `a`'s confirmation until after it has published `b`, since both messages are published in the same EM tick. The only ordering that may flip flop is `'confirm a'`, `'confirm b'`. I think that's only possible if the two messages are routed to different queues. See "When will messages be confirmed?" (https://www.rabbitmq.com/confirms.html).

EventMachine is nice because it provides some guarantees about order, so we don't have to worry about as much. If we were using threads/synchronization I'd have much less hair :)